cupyx/scipy/sparse/index.py

```diff
@@ -0,0 +1,1095 @@
+"""Indexing mixin for sparse matrix classes.
+"""
+import math
+from .sputils import isintlike
+
+from numpy import integer
+from numpy import dtype
+
+import cupy
+import cupyx
+from cupy import core
+
+
+INT_TYPES = (int, integer)
+
+_int32_dtype = dtype('int32')
+_int64_dtype = dtype('int64')
+_float32_dtype = dtype('float32')
+_float64_dtype = dtype('float64')
+_complex64_dtype = dtype('complex64')
+_complex128_dtype = dtype('complex128')
+
+_supported_types = {
+    _int32_dtype: 'int',
+    _int64_dtype: 'long long',
+    _float32_dtype: 'float',
+    _float64_dtype: 'double',
+    _complex64_dtype: 'complex<float>',
+    _complex128_dtype: 'complex<double>'
+}
+
+_module_options = ('--std=c++11',)
+
+
+def _build_name_expressions(types, kernel_name):
+
+    def parse_types(t):
+        if isinstance(types[0], tuple):
+            return ','.join([_supported_types[tu] for tu in t])
+        else:
+            return _supported_types[t]
+
+    return {t: '%s<%s>' % (kernel_name, parse_types(t)) for t in types}
+
+
+_csr_column_index2_order_types = \
+    _build_name_expressions([_int32_dtype, _int64_dtype],
+                            '_csr_column_index2_order')
+_csr_column_index2_order_ker = core.RawModule(code="""
+    template<typename I>
+    __global__
+    void _csr_column_index2_order(const I *col_order,
+                                  const I n,
+                                  I *out_index) {
+
+    // Get the index of the thread
+    I i = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if(i < n)
+        out_index[col_order[i]] = col_order[i+1];
+}
+""", options=_module_options, name_expressions=tuple(
+    _csr_column_index2_order_types.values()
+))
+
+
+def _csr_column_index2_order(col_order, tpb=1024):
+    """Index an array of argsorted indices for looking up
+    locations of duplicates.
+
+    Args
+       col_order : array of argsorted indices from index
+
+    Returns
+        out_order : new array where each element points to
+                    its next potential duplicate in the index
+    """
+    grid = math.ceil(col_order.size-1 / tpb)
+    kernel = _csr_column_index2_order_ker.get_function(
+        _csr_column_index2_order_types[col_order.dtype]
+    )
+    out_order = cupy.empty_like(col_order)
+    kernel((grid,), (tpb,), (col_order, col_order.size-1, out_order))
+
```

Comment:
To keep the maintainability, could you avoid defining custom kernels if not needed? I think most of the kernels in this file can be replaced with existing kernels.
```suggestion
    out_order = cupy.empty_like(col_order)
    out_order[col_order[:-1]] = col_order[1:]
```